### ISSUE TITLE
`ty` fails so find solution in call with invariant containers

### ISSUE THREAD

**Issue by randolf-scholz** [astral-sh/ty#3236]
### Summary

`ty` seems to incorrectly infer `T1=A|B` and `T2=A|B` in the example below, when `T1=A` and `T2=B` is a proper solution. This is a special case of #2615 without slack variables (xref: https://github.com/facebook/pyrefly/issues/2445)

### `main.py`

```py
class Vec[T]:  # invariant in T
    def _items(self) -> list[T]: ...
    def _append(self, item: T) -> None: ...

def concat_vecs[T1, T2](left: Vec[T1], right: Vec[T2]) -> Vec[T1 | T2]:
    return Vec()

class A: ...
class B: ...

def test_vec_concat(left: Vec[A], right: Vec[B]) -> None:
    _0: Vec[A | B] = concat_vecs(left, right) # ❌️
```

```
Argument to function `concat_vecs` is incorrect: Expected `Vec[A | B]`, found `Vec[A]` (invalid-argument-type) [Ln 12, Col 34]
Argument to function `concat_vecs` is incorrect: Expected `Vec[A | B]`, found `Vec[B]` (invalid-argument-type) [Ln 12, Col 40]
```

### `ty.json`

```json
{
    "environment": {
        "python-version": "3.14"
    },
    "rules": {
        "undefined-reveal": "ignore"
    }
}
```

**Comment by randolf-scholz** [comment]
The constraints in the call ``_0: Vec[A | B] = concat_vecs(left, right)`` are:

- `T1 | T2 <: A|B` and `T1 | T2 :> A|B` from outer context and invariance of `Vec`.
- `T1 <: A` and `T1 :> A` via the first argument and invariance of `Vec`.
- `T2 <: B` and `T2 :> B` via the second argument and the invariance of `Vec`.

so the last two sets of constraints should already fix `T1=A` and `T2=B`.

**Comment by carljm** [comment]
Thanks for the report! This is interesting because it's actually just the outer type context that confuses us. If we just have `_0 = concat_vecs(left, right)` we have no issue with the call and we correctly infer `Vec[A | B]`. But when we try to derive additional constraints from matching the type context `Vec[A | B]` against the return type `Vec[T1 | T2]`, we run afoul of #2799.

I think this is basically another case of #2799.
